[Composable API] Add fully_shard debug function to print sharded tree structure, module names and managed param fqns

zhaojuanmao opened the pull request:
Adding a fully_shard debug function to print sharded tree structure like following format, return module names and their managed parameter fqns as well.

![Screenshot 2023-04-18 at 5 14 54 PM](https://user-images.githubusercontent.com/48731194/232931628-169a63a9-b4d5-4902-9cfd-f40113f3ec98.png)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99133

zhaojuanmao committed:
[Composable API] Add two helper functions for fully_shard

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[Composable API] Add two helper functions for fully_shard"

Add two helper functions for 'fully_shard', it will help validating whether sharding is expected and validating parameters managed by fully_shard.




[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -87,6 +89,21 @@ def _get_module_fsdp_state_if_fully_sharded_module(
     return None
 
 
+# Get all the sharded submodule names under the 'root_module' that is usually passed to 'fully_shard()'
```
I want to clarify this comment (and similarly for the other helper).

Is the expectation that the user only passes `module` as the exact module passed to `fully_shard()`? Is this a hard requirement, or is this just the expected usage?

---

My concern with the current implementation is that:
- `_fully_sharded_submodule_names` can be accessed by any submodule managed by `fully_shard()`.
- `_fully_sharded_submodule_names` is not shared across different manual `fully_shard()` calls.

What this means is that a user should be able to call `_get_fully_sharded_submodule_names()` on any submodule managed by `fully_shard()` without erroring but will only see the expected full list of submodule names if using auto wrapping, not manual wrapping.

The reason is that (maybe non-intuitively), only for auto wrapping do submodules managed by `fully_shard()` share the same `_FSDPState` object. The design to share the same `_FSDPState` object is a helpful step for implementing any non-recursive wrapping design, but at the moment, we just have this design difference between wrapper FSDP and `fully_shard()`.

(If you add a unit test that uses manual `fully_shard()` "wrapping", I think the helper will not return the list of names you expect.)

zhaojuanmao replied:
what you are saying is correct, I did have to call the helper twice if I call fully_shard() twice one submodule first and then root module (not manual wrapping in the mode codes, but the two calls are composed/nested).

I think it probably makes more sense to get "_all_fsdp_states" to get all the children's sharding info and managed parameters in this case. as usually I just want to get this debug info in one call

awgu commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -184,6 +201,47 @@ def _is_fsdp_flattened(tensor: torch.Tensor) -> bool:
     return getattr(tensor, FSDP_FLATTENED, False)
 
 
+def _get_handles_to_fqns(
```
Since we are mapping `FlatParameter` to FQN, shall we name this `_get_flat_param_to_fqns`? It looks like above you already named the variable `flat_param_to_fqns`.

zhaojuanmao replied:
_get_flat_param_to_fqns is kind of being used in state dict codes.., and it is designed for non-composable path. Since composable path, there is no 'real' flat_param concept, possibly I can change it to be '_get_flat_param_handles_to_fqns()'? and the key could be handle, not flat_param

awgu replied:
Feel free to disagree, but I feel that even for composable path, there is still the concept of `FlatParameter`, only that it is not exposed to the user.

I think here, the important thing is to have the function name match the `dict`'s keys and values. I think `_get_flat_param_handle_to_fqns()` is fine since that means it maps `flat_param_handle: FlatParamHandle` to `fqns: List[str]`.

awgu commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -184,6 +201,47 @@ def _is_fsdp_flattened(tensor: torch.Tensor) -> bool:
 
+def _get_handles_to_fqns(
+    model: torch.nn.Module,
+) -> Dict[nn.Parameter, List[str]]:
```
nit: if possible
```suggestion
) -> Dict[flat_param_file.FlatParameter, List[str]]:
```

awgu commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -184,6 +201,47 @@ def _is_fsdp_flattened(tensor: torch.Tensor) -> bool:
+    def return_fn(param_to_fqns):
+        return param_to_fqns
+
+    param_to_unflat_param_names: Dict[torch.nn.Parameter, List[str]] = {}
```
nit: I feel like we have some inconsistency in the naming that we should try to cut down on now. Originally, when I first onboarded, I had a mental model of "flat" parameter names (actually `FlatParameter` FQNs) and "unflat" parameter names (actually original parameter FQNs), which is why so much of the optimizer state checkpointing code uses the terms like `unflat_param_names`.

However, since then, I learned you guys use the term FQN, so I started migrating most of the core code to do that, only leaving optimizer state checkpointing not-yet-migrated.

In this case, I feel we should change `return_fn` to use `flat_param_to_fqns` instead of `param_to_fqns`. If there are variable shadowing issues, then you can prefix with `_` for `return_fn()`.

Then, maybe we can change `param_to_unflat_param_names` to also be `flat_param_to_fqns`. This makes it clearer throughout that it is the same quantity (being populated in the `module_fn`, being returned from this helper, and being used downstream in the unit test).

awgu reviewed (commented):
I mainly have some concern about the behavior under manual `fully_shard()` instead of auto wrap. The behavior might not be what you expect, so I wanted to clarify.

zhaojuanmao commented on test/distributed/_composable/fully_shard/test_fully_shard_init.py:
```diff
@@ -7,7 +7,7 @@
 import torch
 import torch.distributed as dist
 import torch.distributed.fsdp._traversal_utils as traversal_utils
-import torch.nn as nn
```
from lint

awgu replied:
Oh interesting. Are you using lintrunner? I have not seen this before. Either I need to update my lintrunner, or maybe you have a different linter?

zhaojuanmao committed:
Update on "[Composable API] Add fully_shard debug function to print sharded tree structure, module names and managed param fqns"

Adding a fully_shard debug function to print sharded tree structure like following format, return module names and their managed parameter fqns as well.

![Screenshot 2023-04-18 at 5 14 54 PM](https://user-images.githubusercontent.com/48731194/232931628-169a63a9-b4d5-4902-9cfd-f40113f3ec98.png)




[ghstack-poisoned]

rohan-varma commented on test/distributed/_composable/fully_shard/test_fully_shard_util.py:
```diff
@@ -0,0 +1,119 @@
+            sharded_module_name_to_fqns,
+        ) = _get_sharded_module_tree_with_module_name_to_fqns(model)
+        if self.rank == 0:
+            print(sharded_tree_info)
```
nit:need this print?

rohan-varma commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -272,13 +274,14 @@ def _apply_to_modules(
             if submodule is None:
                 continue
             new_prefix = prefix + submodule_name + "."
+            new_tree_level = tree_level + 1
```
I might be missing something, but where is tree_level actually used here?

rohan-varma replied:
Oh nvm, this is needed because tree_level is used by the helper.

Should we make it optional so other consumers of apply_to_modules don't have to add it to their API?

zhaojuanmao replied:
it will be hard to make it optional based on current design, prefix is not used by consumers as well. But it is still a required signature, as it is used in the common traversal codes in "def f()"

rohan-varma commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -272,13 +274,14 @@ def _apply_to_modules(
+    def f(module: torch.nn.Module, prefix: str, tree_level: int, *args, **kwargs):
         # Call the module function before recursing over children (pre-order)
-        module_fn(module, prefix, *args, **kwargs)
+        module_fn(module, prefix, tree_level, *args, **kwargs)
```
unrelated to this PR, but maybe module_fn should specify the signature it expects for users of the API.

rohan-varma commented on torch/distributed/fsdp/_debug_utils.py:
```diff
@@ -0,0 +1,95 @@
+                    u2.seq.2[ReLU]
+                u2.l2[Linear]
+            l2[Linear]
+      2. a dict mapping from the concated module FQN and class name to a list of its managed 
```
nit: example of the dict?

rohan-varma reviewed (approved):
thanks

zhaojuanmao committed:
Update on "[Composable API] Add fully_shard debug function to print sharded tree structure, module names and managed param fqns"

Adding a fully_shard debug function to print sharded tree structure like following format, return module names and their managed parameter fqns as well.

![Screenshot 2023-04-18 at 5 14 54 PM](https://user-images.githubusercontent.com/48731194/232931628-169a63a9-b4d5-4902-9cfd-f40113f3ec98.png)




[ghstack-poisoned]

zhaojuanmao commented:
@pytorchbot merge
